src/components/data/_tests/data-utils.test.js

```diff
@@ -428,3 +429,304 @@ describe('prepareAttributeForUpload', () => {
+      const onChange = jest.fn()
+      const { getByLabelText } = render(h(AttributeInput, {
+        value: { entityType: 'thing', entityName: 'thing_one' },
+        onChange
+      }))
+      const valueInput = getByLabelText('New value')
+
+      expect(valueInput.tagName).toBe('INPUT')
+      expect(valueInput.type).toBe('text')
+      expect(valueInput.value).toBe('thing_one')
+
+      fireEvent.change(valueInput, { target: { value: 'thing_two' } })
+      expect(onChange).toHaveBeenCalledWith({ entityType: 'thing', entityName: 'thing_two' })
+    })
+
+    it('renders a number input for number attributes', () => {
+      const onChange = jest.fn()
+      const { getByLabelText } = render(h(AttributeInput, {
+        value: 123,
+        onChange
+      }))
+      const valueInput = getByLabelText('New value')
+
+      expect(valueInput.tagName).toBe('INPUT')
+      expect(valueInput.type).toBe('number')
+      expect(valueInput.value).toBe('123')
+
+      fireEvent.change(valueInput, { target: { value: '456' } })
+      expect(onChange).toHaveBeenCalledWith(456)
+    })
+
+    it('renders a switch for boolean attributes', () => {
+      const onChange = jest.fn()
+      const { getByLabelText } = render(h(AttributeInput, {
+        value: true,
+        onChange
+      }))
+      const valueInput = getByLabelText('New value')
+
+      expect(valueInput.tagName).toBe('INPUT')
+      expect(valueInput.type).toBe('checkbox')
+      expect(valueInput.checked).toBe(true)
+
+      fireEvent.click(valueInput)
+      expect(onChange).toHaveBeenCalledWith(false)
+    })
+  })
+
+  describe('lists', () => {
+    it('renders a checkbox for list status', () => {
+      const { getByLabelText } = render(h(AttributeInput, {
+        value: { itemsType: 'AttributeValue', items: ['foo', 'bar', 'baz'] },
+        onChange: jest.fn()
+      }))
+
+      const listCheckbox = getByLabelText('Value is a list')
+      expect(listCheckbox.getAttribute('role')).toBe('checkbox')
+      expect(listCheckbox.getAttribute('aria-checked')).toBe('true')
+    })
+
+    it('it converts single value to a list when list checkbox is checked', () => {
+      const onChange = jest.fn()
+      const { getByLabelText } = render(h(AttributeInput, {
+        value: 'foo',
+        onChange
+      }))
+
+      const listCheckbox = getByLabelText('Value is a list')
+      fireEvent.click(listCheckbox)
+
+      expect(onChange).toHaveBeenCalledWith({ itemsType: 'AttributeValue', items: ['foo'] })
+    })
+
+    it('it converts list to a single value when list checkbox is unchecked', () => {
+      const onChange = jest.fn()
+      const { getByLabelText } = render(h(AttributeInput, {
+        value: { itemsType: 'AttributeValue', items: ['foo', 'bar', 'baz'] },
+        onChange
+      }))
+
+      const listCheckbox = getByLabelText('Value is a list')
+      fireEvent.click(listCheckbox)
+
+      expect(onChange).toHaveBeenCalledWith('foo')
```

Comment:
Love the clarity of this, but I wonder if the behavior could be surprising.

Reply:
I suppose string lists could be concatenated to convert them to a single value, but for other types I'm not sure if there's another option besides truncation. And for what it's worth, the impact of surprising behavior is somewhat limited since the user can see the new value before saving it to the data table.

Reply:
This comment was unhelpful, sorry. I should have suggested an alternative. If I had tried, I would have discovered that I didn't have any good ideas.